libsrc/apple2/get_tv.s

```diff
@@ -0,0 +1,151 @@
+        .export     _get_tv
+
+        .import     _set_iigs_speed, _get_iigs_speed
+        .import     _bzero, _memset, pushax
+        .import     ostype
+
+        .importzp   tmp1, ptr1
+
+        .constructor calibrate_tv, 2
+
+        .include    "apple2.inc"
+        .include    "accelerator.inc"
+
+.segment "ONCE"
+
+; Cycle wasters
+waste_72:
+        jsr     waste_36
+waste_36:
+        jsr     waste_12
+waste_24:
+        jsr     waste_12
+waste_12:
+        rts
+
+.proc calibrate_tv
+        lda     ostype
+        bmi     iigs
+        cmp     #$20
+        bcc     iip
+        cmp     #$40
+        bcc     iie
+
+iic:    jmp     calibrate_iic
+iigs:   jmp     calibrate_iigs
+iie:    jmp     calibrate_iie
+iip:    rts                   ; Keep TV_OTHER.
+.endproc
+
+; Carry set at enter: wait for VBL +
+; Carry clear at enter: wait for VBL -
+; Increments X every 85 cycles.
+.proc count_until_vbl_bit
+        lda     #$10          ; BPL
+        bcc     :+
+        lda     #$30          ; BMI
+:       sta     sign
+
+        bit     PTRIG         ; Slow accelerators hopefully
+
+        ; Wait for VBLsign change with 85 cycles loops
+:       nop                   ; 2
+        inx                   ; 4
+        jsr     waste_72      ; 78
+        bit     RDVBLBAR      ; 82 - Wait for VBL change
+sign:
+        bpl     :-            ; 85 - patched with bpl/bmi
+        rts
+.endproc
+
+.proc calibrate_iic
+        php
+        sei
+
+        sta     IOUDISOFF
+        lda     RDVBLMSK
+        pha                   ; Back up for cleanup
+
+        bit     ENVBL
+        bit     PTRIG         ; Reset VBL interrupt flag
+:       bit     RDVBLBAR      ; Wait for one VBL
+        bpl     :-
+
+        ldx     #$00
+        clc
+        jsr     count_until_vbl_bit
+
+        pla                   ; Cleanup
+        asl
+        bcs     :+            ; VBL interrupts were already enabled
+        bit     DISVBL
+:       sta     IOUDISON      ; IIc Tech Ref Man: The firmware normally leaves IOUDIS on.
+
+        plp
+        jmp     vbl_calibrate_done
+.endproc
+
+.proc calibrate_iie
+:       bit     RDVBLBAR      ; Wait for bit 7 to be off (VBL start)
+        bmi     :-
+:       bit     RDVBLBAR      ; Wait for bit 7 to be on (VBL end)
+        bpl     :-
+
+        ; Wait and count during a full cycle
+        ldx     #0
+        sec
+        jsr     count_until_vbl_bit
+        clc
+        jsr     count_until_vbl_bit
+
+        jmp     vbl_calibrate_done
+.endproc
+
+.proc calibrate_iigs
+        ; Backup speed and slow down
+        jsr     _get_iigs_speed
+        pha
+        lda     #SPEED_SLOW
+        jsr     _set_iigs_speed
+
+        ; The same as IIe, but reverted, because fuck you that's why
```

Comment:
* I personally dislike any swearing in source code comments.
* Most of the time, something considered stupid turns out to have some valid reason leaving the one calling it stupid behind looking stupid ;-)

Reply:
Sorry! copy-pasted it from my code without thinking twice :-)